src/hotspot/share/opto/vectorization.hpp

```diff
@@ -1323,6 +1323,350 @@ class AlignmentSolver {
+  void apply();
+
+private:
+  // VLoop accessors
+  PhaseIdealLoop* phase()     const { return _vloop.phase(); }
+  PhaseIterGVN& igvn()        const { return _vloop.phase()->igvn(); }
+  IdealLoopTree* lpt()        const { return _vloop.lpt(); }
+  CountedLoopNode* cl()       const { return _vloop.cl(); }
+  PhiNode* iv()               const { return _vloop.iv(); }
+  int iv_stride()             const { return cl()->stride_con(); }
+  bool in_bb(const Node* n)   const { return _vloop.in_bb(n); }
+
+  // VLoopVPointer accessors
+  const VPointer& vpointer(const MemNode* mem) const {
+    return _vloop_analyzer.vpointers().vpointer(mem);
+  }
+
+  void schedule_collect_nodes_without_req_or_dep(GrowableArray<VTransformNode*>& stack) const;
+
+  template<typename Callback>
+  void for_each_memop_in_schedule(Callback callback) const;
+
+  void apply_memops_reordering_with_schedule() const;
+
+  // Ensure that the main loop vectors are aligned by adjusting the pre loop limit.
+  void determine_mem_ref_and_aw_for_main_loop_alignment();
+  void adjust_pre_loop_limit_to_align_main_loop_vectors();
+
+  void apply_vectorization() const;
+
+#ifndef PRODUCT
+  void print_vtnodes() const;
+  void print_schedule() const;
+  void print_memops_schedule() const;
+  void trace_schedule_cycle(const GrowableArray<VTransformNode*>& stack,
+                            const VectorSet& pre_visited,
+                            const VectorSet& post_visited) const;
+#endif
+};
+
+// VTransformNodes resemble the C2 IR Nodes. They represent the resulting scalar and
+// vector nodes as closely as possible.
+class VTransformNode : public ArenaObj {
+public:
+  const VTransformNodeIDX _idx;
+
+private:
+  // _in is split into required inputs (_req), and additional dependencies.
+  const uint _req;
+  GrowableArray<VTransformNode*> _in;
+  GrowableArray<VTransformNode*> _out;
+
+public:
+  VTransformNode(VTransformGraph& graph, const uint req) :
+    _idx(graph.new_idx()),
+    _req(req),
+    _in(graph.arena(),  req, req, nullptr),
+    _out(graph.arena(), 4, 0, nullptr)
+  {
+    graph.add_vtnode(this);
+  }
+
+  void set_req(uint i, VTransformNode* n) {
+    assert(i < _req, "must be a req");
+    assert(_in.at(i) == nullptr && n != nullptr, "only set once");
+    _in.at_put(i, n);
+    n->add_out(this);
+  }
+
+  void swap_req(uint i, uint j) {
+    assert(i < _req, "must be a req");
+    assert(j < _req, "must be a req");
+    VTransformNode* tmp = _in.at(i);
+    _in.at_put(i, _in.at(j));
+    _in.at_put(j, tmp);
+  }
+
+  void add_dependency(VTransformNode* n) {
+    assert(n != nullptr, "no need to add nullptr");
+    _in.push(n);
+    n->add_out(this);
+  }
+
+  void add_out(VTransformNode* n) {
+    _out.push(n);
+  }
+
+  uint req() const { return _req; }
+  VTransformNode* in(int i) const { return _in.at(i); }
+  int outs() const { return _out.length(); }
+  VTransformNode* out(int i) const { return _out.at(i); }
+
+  bool has_req_or_dep() const {
```

Comment:
To match `add_dependency`, I suggest to go with:
```suggestion
  bool has_req_or_dependency() const {
```

Reply:
Good idea.

Reply:
Also renaming
`schedule_collect_nodes_without_req_or_dep -> schedule_collect_nodes_without_req_or_dependency`